### DISCUSSION TITLE
Error with estimate_u_using_random_sampling(target_rows=1e9)

### DISCUSSION THREAD

**Discussion by jkginfinite** [moj-analytical-services/splink#1372] [Q&A]
Hello;

I am currently working with a  200 million row dataset on PySpark EMR. I have 5 blocking rules that generate between 4.3million to 5.8million comparisons each. I am loading the dataset via S3 parquet file to EMR.
I run spark with a driver memory of 1gb and executor memory of 9gb with a total of 4 nodes (it seems to shutdown randomly at higher memories for the driver).

I have encountered an error;
linker.estimate_u_using_random_sampling(target_rows=1e9)
  File "/efs/dq_i/python/aws3-011123/aws3/lib/python3.10/site-packages/splink/linker.py", line 879, in estimate_u_using_random_sampling
    estimate_u_values(self, target_rows)
  File "/efs/dq_i/python/aws3-011123/aws3/lib/python3.10/site-packages/splink/estimate_u.py", line 64, in estimate_u_values
    result = dataframe.as_record_dict()
  File "/efs/dq_i/python/aws3-011123/aws3/lib/python3.10/site-packages/splink/spark/spark_linker.py", line 52, in as_record_dict
    return self.spark_linker.spark.sql(sql).toPandas().to_dict(orient="records")

I am wondering if anyone else has encountered a similar error and what they did about it. Some of the language models suggest it could be a memory error. I have tried changing the blocking rules, etc.

**Comment by RobinL** [comment]
What happens if you reduce target_rows from its current 1 billion to say 10 million?

If the smaller target_rows works that suggests its memory error

**Reply by jkginfinite** [reply]
it returned the same error for 10 million as well as 1 million;

The terminal output did say;

----estimating u probabilities---- before producing the same error as typed above 

at 10million and 1 million (even tried 100k)

**Reply by RobinL** [reply]
Thanks - I think that rules out a memory error.

I think there's some of the error message (i.e. full traceback) missing - it doesn't seem to say what the error actually is.  Possibly because the error is occurring on the Spark side? - i.e. the cause of the error is potentially in the Spark logs.  Are you able to retrieve any more information about the error?  

One option would be to run the code on a small sample of data locally to see if you get the same error.  In terms of our workflow for developing linkage jobs, we generally start that way with Spark - i.e. get something running locally on a small sample of records, and only deploy to the cluster when we're sure there aren't any typos etc. in our scripts

**Reply by jkginfinite** [reply]
Hey Robin, 

I have been able to train on small datasets (100k) but beyond that it gives this error (above)
still getting this error;


Do you think this could have been caused by Splink DuckDB hack we did months ago? 

here's a link to that discussion; https://github.com/moj-analytical-services/splink/discussions/1239

**Reply by RobinL** [reply]
It's a little peculiar because the estimate u step should not usually be affected by the size of the input dataset. The first thing it does it take a smallish random sample from the input dataset, before continuing data processing on that random sample.

Instead, the performance is mostly determined by the target_rows.  

I'm also a little confused because i believe the error is being generated from [this line](https://github.com/moj-analytical-services/splink/blob/606a99d61e6502511210b70c7db0873cbc231956/splink/estimate_u.py#L62) of code, which is just running a simple count(*).  Incidentally, version of Splink are you running?  That will allow me to verify it is that line that's causing the problem

How are you loading data in?  Is it possible it's coming in from a single (or very small number) of huge parquet files?  If so, you will want to run e.g. a  `df = df.repartition(100)` prior to loading the data into Splink.  That's might be a good idea anyway, just in case

Do you have a lot of term frequency adjustments turned on?  That could be a possible cause since the estimate u step computes all term frequency tables.  They're not actually strictly needed for estimate u, but they're cached for later.  You could try turning them off.

Another possible cause is if you're running a `link_type = "link_only"` job with very unbalanced datasets (e.g. some with very few rows, other with loads).  A bug with that has [just been fixed in the 3.9.3 release](https://github.com/moj-analytical-services/splink/pull/1359).

Are you able to share your script? (Feel free to change variable names if they're sensitive).  Either post here or could sent to robin.linacre@digital.justice.gov.uk

Also, any additional information on the error would be useful (feel free to email).  For example, Spark logs.  Or if you can collect the logs produced by:
import logging 
logging.getLogger("splink").setLevel(1)
that would be super useful

**Reply by RobinL** [reply]
One more thing: just to check you're setting your spark checkpoint directory to somewhere in s3 (or HDDs) rather than a local directory

**Reply by jkginfinite** [reply]
Hi,

We're using version 3.6.0 for splink.

So your saying have the spark context check point to be a path in S3? 

previously our path structure was "hdfs://<active-namenode-hostname>:8020/user/user_name/cache" for the checkpoint

Since I had the break lineage method set to "parquet"

I will try checkpoint = "s3://some path in s3"

I am loading in the form of data = spark.read.parquet(s3_file_path_to_parquet)

I can email you the script as well;

**Reply by jkginfinite** [reply]
@RobinL sent you an email :)

**Reply by RobinL** [reply]
Thanks - will reply via email separately but re: the s3 question, I think your current settings using hdfs are fine, but you might try S3 just to verify that's not the source of the problem
